File: packages/altimate-engine/src/altimate_engine/connectors/postgres.py
from __future__ import annotations

from typing import Any

from altimate_engine.connectors.base import Connector


class PostgresConnector(Connector):
    def __init__(
        self,
        host: str = "localhost",
        port: int = 5432,
        database: str = "postgres",
        user: str | None = None,
        password: str | None = None,
        connection_string: str | None = None,
        **kwargs,
    ):
        self.connection_string = connection_string
        self.host = host
        self.port = port
        self.database = database
        self.user = user
        self.password = password
        self.kwargs = kwargs
        self._conn = None

    def connect(self) -> Any:
        try:
            import psycopg2
        except ImportError:
            raise ImportError(
                "psycopg2 not installed. Install with: pip install altimate-engine[warehouses]"
            )

        if self.connection_string:
            self._conn = psycopg2.connect(self.connection_string)
        else:
            self._conn = psycopg2.connect(
                host=self.host,
                port=self.port,
                database=self.database,
                user=self.user,
                password=self.password,
                **self.kwargs,
            )
        return self._conn

    def _ensure_conn(self):
        if self._conn is None:
            self.connect()
        return self._conn

    def execute(self, sql: str, params: tuple | list | None = None, limit: int = 1000) -> list[dict[str, Any]]:
        conn = self._ensure_conn()
        cur = conn.cursor()
        cur.execute(sql, params)

        if cur.description is None:
            conn.commit()
            return []

        columns = [desc[0] for desc in cur.description]
        rows = cur.fetchmany(limit)
        result = [dict(zip(columns, row)) for row in rows]
        cur.close()
        return result

    def list_schemas(self) -> list[str]:
        rows = self.execute("SELECT schema_name FROM information_schema.schemata")
        return [row["schema_name"] for row in rows]

    def list_tables(self, schema: str) -> list[dict[str, Any]]:
        rows = self.execute(
            "SELECT table_name, table_type FROM information_schema.tables WHERE table_schema = %s",
            (schema,),
        )
        return [{"name": row["table_name"], "type": row["table_type"]} for row in rows]

    def describe_table(self, schema: str, table: str) -> list[dict[str, Any]]:
        rows = self.execute(
            "SELECT column_name, data_type, is_nullable FROM information_schema.columns WHERE table_schema = %s AND table_name = %s",
            (schema, table),
        )
        return [
            {
                "name": row["column_name"],
                "data_type": row["data_type"],
                "nullable": row["is_nullable"] == "YES",
            }
            for row in rows
        ]

    def set_statement_timeout(self, timeout_ms: int) -> None:
        """Set PostgreSQL session statement timeout.

        Args:
            timeout_ms: Maximum query execution time in milliseconds.
        """
        conn = self._ensure_conn()
        cur = conn.cursor()
        cur.execute(f"SET statement_timeout = {int(timeout_ms)}")
        conn.commit()
        cur.close()

    def close(self) -> None:
        if self._conn is not None:
            self._conn.close()
            self._conn = None

File: packages/drivers/src/postgres.ts
/**
 * PostgreSQL driver using the `pg` package.
 */

import type { ConnectionConfig, Connector, ConnectorResult, SchemaColumn } from "./types"

export async function connect(config: ConnectionConfig): Promise<Connector> {
  let pg: any
  try {
    pg = await import("pg")
  } catch {
    throw new Error("PostgreSQL driver not installed. Run: npm install pg @types/pg")
  }

  const Pool = pg.default?.Pool ?? pg.Pool
  let pool: any

  const connector: Connector = {
    async connect() {
      const poolConfig: Record<string, unknown> = {}

      if (config.connection_string) {
        poolConfig.connectionString = config.connection_string
      } else {
        // Validate required credentials before connecting to avoid cryptic
        // SASL/SCRAM errors from the pg driver when password is missing
        if (config.password != null && typeof config.password !== "string") {
          throw new Error(
            "PostgreSQL password must be a string. Check your warehouse configuration.",
          )
        }
        poolConfig.host = config.host ?? "127.0.0.1"
        poolConfig.port = config.port ?? 5432
        poolConfig.database = config.database ?? "postgres"
        poolConfig.user = config.user
        poolConfig.password = config.password
        if (config.ssl !== undefined) {
          poolConfig.ssl = config.ssl
        }
      }

      poolConfig.max = 5
      poolConfig.idleTimeoutMillis = 30000
      poolConfig.connectionTimeoutMillis = 10000

      pool = new Pool(poolConfig)
    },

    async execute(sql: string, limit?: number, _binds?: any[]): Promise<ConnectorResult> {
      const client = await pool.connect()
      try {
        if (config.statement_timeout) {
          const timeoutMs = Number(config.statement_timeout)
          if (Number.isFinite(timeoutMs) && timeoutMs > 0) {
            await client.query(`SET statement_timeout TO ${Math.round(timeoutMs)}`)
          }
        }

        let query = sql
        const effectiveLimit = limit ?? 1000
        const isSelectLike = /^\s*(SELECT|WITH|VALUES)\b/i.test(sql)
        // Add LIMIT only for SELECT-like queries and if not already present
        if (
          isSelectLike &&
          effectiveLimit &&
          !/\bLIMIT\b/i.test(sql)
        ) {
          query = `${sql.replace(/;\s*$/, "")} LIMIT ${effectiveLimit + 1}`
        }

        const result = await client.query(query)
        const columns = result.fields?.map((f: any) => f.name) ?? []
        const truncated = result.rows.length > effectiveLimit
        const rows = truncated
          ? result.rows.slice(0, effectiveLimit)
          : result.rows

        return {
          columns,
          rows: rows.map((row: any) => columns.map((col: string) => row[col])),
          row_count: rows.length,
          truncated,
        }
      } finally {
        client.release()
      }
    },

    async listSchemas(): Promise<string[]> {
      const result = await connector.execute(
        `SELECT schema_name FROM information_schema.schemata
         WHERE schema_name NOT IN ('information_schema', 'pg_catalog', 'pg_toast')
         ORDER BY schema_name`,
        10000,
      )
      return result.rows.map((r) => r[0] as string)
    },

    async listTables(
      schema: string,
    ): Promise<Array<{ name: string; type: string }>> {
      const client = await pool.connect()
      try {
        const result = await client.query(
          `SELECT table_name, table_type
           FROM information_schema.tables
           WHERE table_schema = $1
           ORDER BY table_name`,
          [schema],
        )
        return result.rows.map((r: any) => ({
          name: r.table_name as string,
          type: r.table_type === "VIEW" ? "view" : "table",
        }))
      } finally {
        client.release()
      }
    },

    async describeTable(
      schema: string,
      table: string,
    ): Promise<SchemaColumn[]> {
      const client = await pool.connect()
      try {
        const result = await client.query(
          `SELECT column_name, data_type, is_nullable
           FROM information_schema.columns
           WHERE table_schema = $1
             AND table_name = $2
           ORDER BY ordinal_position`,
          [schema, table],
        )
        return result.rows.map((r: any) => ({
          name: r.column_name as string,
          data_type: r.data_type as string,
          nullable: r.is_nullable === "YES",
        }))
      } finally {
        client.release()
      }
    },

    async close() {
      if (pool) {
        await pool.end()
        pool = null
      }
    },
  }
  return connector
}
